sdk/agent/agent.go

```diff
@@ -52,21 +53,23 @@ type Agent struct {
 
 	Events chan<- Event
 
+	// mu is a lock for the mutable fields of this type. It should be locked
+	// when reading or writing any of the mutable fields. The mutable fields are
+	// listed below. If pushing to a chan, such as Events, it is unnecessary to
+	// lock.
+	mu sync.Mutex
```

Comment:
❓ more for my own understanding but was putting the mutex lock on the channel struct considered? Since the channel is closer to the data that has the data race. Or is it bc we needed to lock the `agent.conn` as well

reason I'm wondering is because if it was on the channel then we could put the responsibility of locking in the sdk eg. on the `UpdateLocalEscrowAccountBalance` and `UpdateRemoteEscrowAccountBalance` methods

Reply:
We could, there were a couple reasons I went with in the agent though.

I'm not 100% sure if it would actually be a problem, but we need to be careful to avoid deadlock whenever there are multiple synchronizing primitives in use. The `conn`, if it's a `net.Conn`, has an internal lock, and the state.Channel will have a lock, and if both are locked in alternating orders we could end up with deadlock. It's a little safer to nest the mutex around all the entities we need to synchronize on.

Another reason is that the state.Channel has no use for a mutex itself. It doesn't run any go routines. So it seems to me like the requirements of the agent would be leaking into the channel. This would mean any use of the `state.Channel` without the agent would be bound to use the synchronizing pattern that we added for the agent, so it would couple the two together a little.

Let me know if you see tradeoffs I'm not seeing.

Reply:
nope I was more just curious than anything. Both those reasons make sense, thx for explaining!